CODING.md

```diff
@@ -0,0 +1,59 @@
+
+## Testing
+
+It is preferred to use separate test package to write tests. This allows explicit testing on exported behaviour of the package, not the internal implementation. In need to export only to testing package, export_test.go file should be used. Testing internal implementations discouraged, but not forbidden. Executing `make test` should pass without any warnings or errors.
+
+## Packages
+
+Go packages with the Bee project should have a single and well defined responsibility, clear exposed API and tests that cover expected behaviour. To ensure better modularity of the codebase, every package should be treated as a library that provides a specific functionality and that can be used as a module in other applications. Every package should have a well written godoc page which should be used as the entry point for understanding the package's responsibility. The same as using any other third-party package. If the package godoc is not clear and requires looking at the code to understand behaviour that it provides, documentation should be improved.
+
+## Concurrency
+
+Handling concurrency in Go is not a trivial task.
+
+For every goroutine that is created it must be defined how and when it terminates.
+
+Every channel must have an owning goroutine. That goroutine is the only one that can send to the channel, close it or transfer ownership to another goroutine. Wherever is possible a readonly or writeonly channels should be used.
+
+## Error propagation, annotation and handling
+
+Errors must be propagated by the caller function. Error should not be logged and passed at the same time. It is up to the next caller function to decide what should be done in case of an error.
+
+If a function has multiple returns with, any returned error should be annotated either with fmt.Errorf using `%w` verb or with custom error type that has Unwrap method.
+
+## Logging
+
+Log messages are categorized in five different severities: error, warning, info, debug and trace, with semantic meaning associated with each of them.
+
+Two types of application users are identified: regular user and developer. Regular user should not be presented with confusing technical implementation details in log messages, but only with meaningful information related to application operability in a form of meaningful statements. Developers are users that are aware of implementation details and they will benefit from technical details to help them debug the application problematic state.
+
+This means that the same problematic event may have two log lines but with different severities. This is the case with Error/Debug or Warning/Debug combo where Error or Warning is meant for the regular user and the Debug for developers to help them investigate the issue. Info and Trace log levels are informative about the expected state changes, but Info, with operable information, for regular user, and Trace, with technical details, for developer.
+
```

Comment:
maybe also nice to have - do we have the annotation on which line the log has been logged from? we had this with the go-ethereum logger, not sure if this one supports this too. if not - the user of the logging library should make sure that the logged message can be traced back to a specific line in a file

Reply:
I think that it is possible with logrus. I am more in favour that our log messages are written in a way that are obvious where they could be found by their meaning, but yes, sometimes it is helpful to have a file and line number. I am of the opinion that this is a sign of not so well written log line.

Reply:
true. it might still be good to mention if the feature exists in the logger, i think this is good for onboarded devs to know

Reply:
I agree. We may add the option to bee cli, and document here that is it encouraged only to be used in development.